### DISCUSSION TITLE
Nested Array type inspection, best practices

### DISCUSSION THREAD

**Discussion by gipert** [scikit-hep/awkward#3049] [Q&A] [answered]
I'm working on a routine that accepts only a specific type of Array, e.g. 
```pycon
>>> ak.Array([[[1, 2], [3]], [[7], []], [[9], [3, 8, 5]]])
<Array [[[1, 2], [3]], ..., [[9], [3, ...]]] type='3 * var * var * int64'>
```
strictly characterized by:

* generic dimensionality
* fully jagged
* non-record-type
* non-option-type

each of these features must be present at each axis.

I'm looking for a way to check the array sanity (i.e. the presence of all these characteristics) at all depths at the same time. I presume one needs to recursively walk the array structure? If yes, what is the recommended way to do so?

**Comment by agoose77** [comment, marked as answer]
Awkward Arrays have three levels of type information:
1. Types ("this is a ragged list", "this is a regular dimension", "this is an option", ...)
2. Forms ("this is a ragged list with starts and stops", ...)
3. Contents ("this is a ragged list with starts and stops whose values are ...", ...)

It sounds like you only care about the first one, `ak.Array.type`.

We would write this kind of test using recursion, e.g.
```python
import awkward as ak


def assert_valid(type_):
    if isinstance(type_, ak.types.ArrayType):
        return assert_valid(type_.content)
    elif isinstance(type_, ak.types.ScalarType):
        raise TypeError("Expected ArrayType or its content")

    assert not isinstance(
        type_,
        (
            ak.types.RegularType,
            ak.types.OptionType,
            ak.types.UnionType,
            ak.types.UnknownType,
            ak.types.RecordType,
        ),
    )
    if isinstance(type_, ak.types.ListType):
        assert_valid(type_.content)
    else:
        assert isinstance(type_, ak.types.NumpyType)


array = ak.Array([[1, 2, 3]])
assert_valid(array.type)

```

These type objects form a closed set, so we can exhaustively test each one. You should take a look at the type object of an array to see what its properties are!


Does this enforce the rules you care about?

**Reply by gipert** [reply]
Seems so thanks!

Just as a clarification, why wouldn't this:
```py
def assert_valid(type_):
    if isinstance(type_, ak.types.ArrayType | ak.types.ListType):
        return assert_valid(type_.content)
    elif isinstance(type_, ak.types.ScalarType):
        raise TypeError("Expected ArrayType or its content")

    assert not isinstance(
        type_,
        (
            ak.types.RegularType,
            ak.types.OptionType,
            ak.types.UnionType,
            ak.types.UnknownType,
            ak.types.RecordType,
            ak.types.NumpyType,
        ),
    )
```
be equivalent to your function?

**Reply by agoose77** [reply]
It would be! The way I wrote the function reflects my thinking on our type hierarchy: `ArrayType` and `ScalarType` are top-level types that shouldn't be seen inside a type tree.
